### PR TITLE
1157 validate percent values are hex

### LINKED ISSUE #1157: URL parsing should validate percent are hex
We should validate percent values are valid hex, for example `%ZZ` should not be valid.

### PR DESCRIPTION
**Pull request by marschall**
Fixes #1157

### CODE REVIEW

File: repository/Seaside-Core.package/WAUrl.class/class/decodePercent..st [hunk 1]
```diff
@@ -12,8 +12,8 @@ decodePercent: aString
 				char = $%
 					ifTrue: [
 						| firstByte secondByte |
-						firstByte :=  input atEnd ifTrue: [ WAInvalidUrlSyntaxError signal: aString ] ifFalse: [ input next asUppercase digitValue ].
-						secondByte :=  input atEnd ifTrue: [ WAInvalidUrlSyntaxError signal: aString ] ifFalse: [ input next asUppercase digitValue ].
+						firstByte :=  self readHexFrom: input errorDetail: aString.
+						secondByte :=  self readHexFrom: input errorDetail: aString.
 						Character codePoint:firstByte * 16 + secondByte ]
```
Comment (theseion) [inline repository/Seaside-Core.package/WAUrl.class/class/decodePercent..st]: Missing space

File: repository/Seaside-Core.package/WAUrl.class/class/readHexFrom.errorDetail..st [hunk 2]
```diff
@@ -0,0 +1,10 @@
+private
+readHexFrom: aStream errorDetail: aString
+	| hexChar |
+	aStream atEnd ifTrue: [
+		WAInvalidUrlSyntaxError signal: aString ].
+	hexChar := aStream next.
```
Comment (theseion) [inline repository/Seaside-Core.package/WAUrl.class/class/readHexFrom.errorDetail..st]: Would bit shift operations make sense here to improve performance?

Comment (marschall) [reply repository/Seaside-Core.package/WAUrl.class/class/readHexFrom.errorDetail..st]: The temptation to optimize is there but I think it's unlikely this shows up during profiling.

**Comment by marianopeck** [comment]
Hi @marschall @theseion @gcotelli 

I was reviewing this old issue and I see you modified `#decodePercent:` to delegate to ` #readHexFrom:errorDetail:`
My question is whether we should actually do the same for `#decodePercent:from:to:` and `#decodePercentAndPlus:from:to:`

thanks in advance

**Comment by gcotelli** [comment]
IMHO yes. Probably nobody noticed the problem with the other methods. Do you want to provide a PR and add some test cases? Looks that both methods are used during parsing.

**Comment by theseion** [comment]
I agree.

**Pull request merged**